src/board.rs

```diff
@@ -0,0 +1,1021 @@
+                }
+            }
+        }
+    }
+
+    pub fn get_root(&self, player: usize) -> Result<Coordinate, &str> {
+        if player >= self.roots.len() {
+            Err("Invalid player")
+        } else {
+            Ok(self.roots[player])
+        }
+    }
+
+    pub fn neighbouring_squares(&self, position: Coordinate) -> Vec<(Coordinate, Square)> {
+        let mut neighbours = Vec::new();
+        for delta in Direction::iter() {
+            let neighbour_coordinate = delta.add(position);
+            match self.get(neighbour_coordinate) {
+                Err(_) => {
+                    continue; // Skips invalid squares
+                }
+                Ok(square) => {
+                    neighbours.push((neighbour_coordinate, square));
+                }
+            }
+        }
+        neighbours
+    }
+
+    // TODO: return iterator or rename since it doesn't matter that this is depth first when we return a HashSet
+    fn depth_first_search(&self, position: Coordinate) -> HashSet<Coordinate> {
+        let mut set = HashSet::new();
+
+        let player = if let Ok(Square::Occupied(player, _)) = self.get(position) {
+            player
+        } else {
+            return set;
+        };
+        let mut stack = vec![position]; // TODO: consider more efficient stack type
+
+        while let Some(current) = stack.pop() {
+            set.insert(current);
+            for (position, square) in self.neighbouring_squares(current) {
+                // Put the neighbour in the set if it is occupied by the current player
+                if let Square::Occupied(neighbours_player, _) = square {
+                    if !set.contains(&position) && player == neighbours_player {
+                        stack.push(position);
+                    }
+                }
+            }
+        }
+
+        set
+    }
+
+    pub fn swap(&mut self, player: usize, positions: [Coordinate; 2]) -> Result<(), &str> {
+        if positions[0] == positions[1] {
+            return Err("Can't swap a square with itself");
+        }
+
+        let mut tiles = ['&'; 2];
+        for (i, pos) in positions.iter().enumerate() {
+            match self.get(*pos) {
+                // TODO: use ? and (possibly) combine function into single match post Polonius
+                Err(_) => return Err("Invalid swap position"),
+                Ok(square) => match square {
+                    Square::Empty => return Err("Must swap between occupied squares"),
+                    Square::Occupied(owner, tile) => {
+                        if owner != player {
+                            return Err("Player must own the squares they swap");
+                        }
+                        tiles[i] = tile;
+                    }
+                },
+            };
+        }
+
+        // TODO: use ? post Polonius
+        if self.set(positions[0], player, tiles[1]).is_err() {
+            return Err("Can't set");
+        }
```

Comment:
I think moving to anyhow as mentioned above will let you use `?` here — any Error type should coerce to an `anyhow::Error` correctly (unless there's a subtlety I'm overlooking)

Reply:
Nope you were totally right! Just pushed some custom errors.